platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ConfigurationCacheFingerprintChecker.kt

```diff
@@ -348,14 +358,29 @@ class ConfigurationCacheFingerprintChecker(private val host: Host) {
         host.displayNameOf(file)
 
     private
-    fun buildLogicInputHasChanged(valueSource: ValueSource<Any, ValueSourceParameters>): InvalidationReason =
+    fun buildLogicInputHasChanged(valueSource: ValueSource<Any, ValueSourceParameters>): InvalidationReason = StructuredMessage.forText(
         (valueSource as? Describable)?.let {
             it.displayName + " has changed"
         } ?: "a build logic input of type '${unpackType(valueSource).simpleName}' has changed"
+    )
 
     private
-    fun buildLogicInputFailed(obtainedValue: ObtainedValue, failure: Throwable): InvalidationReason =
+    fun buildLogicInputFailed(obtainedValue: ObtainedValue, failure: Throwable): InvalidationReason = StructuredMessage.forText(
         "a build logic input of type '${obtainedValue.valueSourceType.simpleName}' failed when storing the entry with $failure"
+    )
+
+    /**
+     * Builds a structured message with a given [block], but if null is returned from the block, discards the message.
+     * @return built message or null if [block] returns null
+     */
+    private
+    inline fun structuredMessageOrNull(block: StructuredMessage.Builder.() -> StructuredMessage.Builder?): StructuredMessage? =
+        StructuredMessage.Builder().run { block() }?.build()
```

Comment:
This intentionally does not use `StructuredMessage.build {}`.

The idea here is to allow discarding the message (returning null), if the `block` we are running tells us so by itself returning the null. This is enforced by having `block` return `StructuredMessage.Builder?`. 

This has an additional nice property that if forces all the `when` branches to either use builder methods like `text()` or `reference()` or explicitly return null.

Using the builder as a receiver of the `block` allows to directly use builder functions inside the block, thus minimizing the diff for all `when` branches.